packages/runtime/runtime-utils/src/dataStoreHelpers.ts

```diff
@@ -36,29 +39,35 @@ export function exceptionToResponse(err: any): IResponse {
         };
     }
 
-    // Capture error objects, not stack itself, as stack retrieval is very expensive operation, so we delay it
-    const errWithStack = generateErrorWithStack();
-
+    const normalizedError = normalizeError(err);
     return {
         mimeType: "text/plain",
-        status,
-        value: `${err}`,
-        get stack() { return ((err?.stack) as (string | undefined)) ?? errWithStack.stack; },
+        status: 500,
+        value: normalizedError.message,
+        get stack() { return normalizedError.stack; },
     };
 }
 
-export function responseToException(response: IResponse, request: IRequest): Error {
-    const message = response.value;
-    const errWithStack = generateErrorWithStack();
-    const responseErr: Error & IResponseException = {
-        errorFromRequestFluidObject: true,
-        message,
-        name: "Error",
-        code: response.status,
-        get stack() { return response.stack ?? errWithStack.stack; },
-    };
+/**
+ * Error class for error instances generated from an error IResponse
+ */
+class ResponseException extends LoggingError implements IResponseException {
+    readonly errorFromRequestFluidObject: true = true;
+    readonly code: number;
+    constructor(response: IResponse) {
+        const responseValue = response.value;
+        const message = typeof responseValue === "string" ? responseValue : String(responseValue);
+        super(message);
+
+        this.code = response.status;
+        if (response.stack !== undefined) {
+            this.overwriteStack(response.stack);
+        }
```

Comment:
Instead of this, I need to add it as a getter to avoid invoking `.stack` prematurely

Reply:
Yes, I remember I had to make some fixes to avoid getting stack all the time - it was very expensive.
I worry that normalizeError() call in exceptionToResponse() adds same problem and we will need to come back and revert it.
Can you please glance at history? I remember Loop provided us that feedback, and you signed off on my perf fix here :)

Reply:
Yep I remember that's how I realized my mistake. I will revisit this and scrutinize it against that original change.